File: report/src/main/java/com/vmlens/report/description/DescriptionContextImpl.java
package com.vmlens.report.description;


import com.vmlens.report.container.ContainerForField;
import com.vmlens.report.container.ContainerForMethod;
import com.vmlens.report.container.ContainerForTestLoop;
import com.vmlens.report.container.ContainerForThread;
import com.vmlens.report.input.LoopRunAndThreadIndex;

import java.util.Map;

public class DescriptionContextImpl implements DescriptionContext {

    private final ContainerMapAdapter<LoopRunAndThreadIndex, ContainerForThread> threadNames;
    private final ContainerMapAdapter<Integer, ContainerForMethod> methodNames;
    private final ContainerMapAdapter<Integer, ContainerForField> fieldNames;
    private final ContainerMapAdapter<Integer, ContainerForTestLoop> loopNames;

    public DescriptionContextImpl(Map<LoopRunAndThreadIndex, ContainerForThread> indexToThreadDescription,
                                  Map<Integer, ContainerForMethod> idToMethodContainer,
                                  Map<Integer, ContainerForField> idToFieldContainer,
                                  Map<Integer, ContainerForTestLoop> idToTestLoopDescription) {
        this.threadNames = new ContainerMapAdapter<>(indexToThreadDescription);
        this.methodNames = new ContainerMapAdapter<>(idToMethodContainer);
        this.fieldNames = new ContainerMapAdapter<>(idToFieldContainer);
        this.loopNames = new ContainerMapAdapter<>(idToTestLoopDescription);
    }

    @Override
    public String threadName(LoopRunAndThreadIndex key) {
        return threadNames.getName(key);
    }

    public String methodName(Integer key) {
        return methodNames.getName(key);
    }

    public String fieldName(Integer key) {
        return fieldNames.getName(key);
    }

    public String loopName(Integer key) {
        return loopNames.getName(key);
    }

}

File: sync-bug/src/main/scala/com/anarsoft/race/detection/report/description/DescriptionContextImpl.scala
package com.anarsoft.race.detection.report.description

import com.anarsoft.race.detection.report.element.LoopRunAndThreadIndex
import com.vmlens.api.AllInterleavingsBuilder
import com.vmlens.trace.agent.bootstrap.description.{ClassDescription, MethodDescription}

class DescriptionContextImpl(val containerMapCollection : ContainerMapCollection) extends  DescriptionContext  {

  override def threadName(key: LoopRunAndThreadIndex): String = {
    containerMapCollection.threadNames.get(key) match {
      case None => {
        notFound(key);
      }
      case Some(x) => x.threadName();
      
    }
  }

  override def methodName(key: Int): String = {
    containerMapCollection.methodNames.get(key).flatMap(  x => x.description )
      .map( x => createMethodString(x)) match {
      case None => {
        notFound(key);
      }
      case Some(x) => x
    }
  }

  override def classNameForMethodId(key: Int): String =  {
    containerMapCollection.methodNames.get(key).flatMap(  x => x.description )
      .map( x => createClassName(x)) match {
      case None => {
        notFound(key);
      }
      case Some(x) => x
    }
  }

  override def methodNameWithoutSource(key: Int): String = {
    containerMapCollection.methodNames.get(key).flatMap(  x => x.description )
      .map( x =>  x._1.name().replace('/', '.') + "." + x._2.name() ) match {
      case None => {
        notFound(key);
      }
      case Some(x) => x
    }
  }

  override def fieldName(key: Int): String = {
    containerMapCollection.fieldNames.get(key).flatMap(  x => x.description )
      .map( x => x._1.name().replace('/','.') + "." + x._2.name() ) match {
      case None => {
        notFound(key);
      }
      case Some(x) => x
    }
  }

  override def loopName(key: Int): String =  {
    containerMapCollection.loopNames.get(key) match {
      case None => {
        notFound(key);
      }
      case Some(x) => x.name();

    }
  }

  override def reportAsSummaryThreshold(loopId: Int): Int = {
    containerMapCollection.loopNames.get(loopId) match {
      case None => {
        AllInterleavingsBuilder.REPORT_AS_SUMMARY_THRESHOLD;
      }
      case Some(x) => x.reportAsSummaryThreshold();

    }
  }

  private def createClassName(tuple: Tuple2[ClassDescription, MethodDescription]): String = {
    tuple._1.name().replace('/', '.') 
  }
  
  
  /*
     * see java.lang.StackTraceElement
     */
  private def createMethodString(tuple : Tuple2[ClassDescription, MethodDescription]): String = {
    tuple._1.name().replace('/', '.') + "." + tuple._2.name() + "("  + classSource(tuple._1.source())
       + lineNumber(tuple._2.lineNumber()) + ")"
  }

  private def classSource(source : String): String = {
    if(source.equals("")) {
      "Unknown Source"
    } else {
      source
    }
  }

  private def lineNumber(number : Int): String = {
    if( number < 0 ) {
      ""
    } else {
      ":" + number;
    }
  }

  private def notFound(key: Any) = String.format("not found (%s)", key)
  
  
  
  
}
